server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -485,11 +486,11 @@ private void recoverFromTranslogInternal(TranslogRecoveryRunner translogRecovery
         // note: if opsRecovered == 0 and we have older translogs it means they are corrupted or 0 length.
         assert pendingTranslogRecovery.get() : "translogRecovery is not pending but should be";
         pendingTranslogRecovery.set(false); // we are good - now we can commit
+        logger.trace(() -> new ParameterizedMessage(
+                "flushing post recovery from translog: ops recovered [{}], current translog generation [{}]",
+                opsRecovered, translog.currentFileGeneration()));
+        flush(false, true);
         if (opsRecovered > 0) {
```

Comment:
I kept this condition around the external refresh triggered here. Not sure if that optimization is necessary now to be honest, but removing it breaks a large number of unit tests that would now see two refreshes instead of one (since the flush also refreshes)

Reply:
The refresh in the flush method is an internal refresh, not an external one. I don't think that the external refresh is actually necessary (we do the relevant refresh in IndexShard.postRecovery, which is called before moving the shard to started). What happens if you just remove the refresh call here?

Reply:
It breaks a few tests that relied on the external refresh when recovering from translog, but it's a manageable number. I'll adjust those and run a few iterations of tests to make sure I didn't miss a spot.